File: src/data/catalog.ts
import type { CategoryKey } from './events';
import {
  getCatalogComingSoon,
  getCatalogComingSoonById,
  getCatalogEventById,
  getCatalogEventsByCategory,
  getCatalogMovieById,
  getCatalogMovies,
} from '../lib/catalogApi';

export interface ShowItem {
  id: string;
  kind: 'movie' | 'event';
  title: string;
  originalTitle: string;
  imageUrl: string;
  ageRating: string;
  formats: string;
  rating: number;
  nextSessionDate: string;
  duration: string;
  genres: string;
  description: string;
  releaseDate: string;
  venue?: string;
  priceFrom?: number;
  priceTo?: number;
}

const CATEGORY_LABELS: Record<string, string> = {
  concerts: 'Concert',
  theater: 'Theater',
  kids: 'Family',
  standup: 'Stand-Up',
};

function movieToShowItem(m: {
  id: string;
  title: string;
  originalTitle: string;
  imageUrl: string;
  ageRating: string;
  formats: string;
  rating: number;
  nextSessionDate: string;
  duration: string;
  genres: string;
  description: string;
  releaseDate: string;
}): ShowItem {
  return {
    id: m.id,
    kind: 'movie',
    title: m.title,
    originalTitle: m.originalTitle,
    imageUrl: m.imageUrl,
    ageRating: m.ageRating,
    formats: m.formats,
    rating: m.rating,
    nextSessionDate: m.nextSessionDate,
    duration: m.duration,
    genres: m.genres,
    description: m.description,
    releaseDate: m.releaseDate,
  };
}

function eventToShowItem(e: {
  id: string;
  category: string;
  title: string;
  venue: string;
  date: string;
  time: string;
  priceFrom: number;
  priceTo: number;
  imageUrl: string;
  description?: string;
}): ShowItem {
  const label = CATEGORY_LABELS[e.category] ?? 'Event';
  return {
    id: e.id,
    kind: 'event',
    title: e.title,
    originalTitle: e.title,
    imageUrl: e.imageUrl,
    ageRating: e.category === 'kids' ? 'All Ages' : '16+',
    formats: label,
    rating: 8.0 + (e.id.charCodeAt(1) % 10) / 10,
    nextSessionDate: `${e.date.toUpperCase()} · ${e.time}`,
    duration: e.category === 'theater' ? '2h 30m' : '2h',
    genres: label,
    description:
      e.description ||
      `${e.title} — an unforgettable ${label.toLowerCase()} experience at ${e.venue}.`,
    releaseDate: `${e.date} · ${e.time}`,
    venue: e.venue,
    priceFrom: e.priceFrom,
    priceTo: e.priceTo,
  };
}

function comingSoonToShowItem(item: {
  id: string;
  kind: 'movie' | 'event';
  title: string;
  originalTitle: string;
  imageUrl: string;
  ageRating: string;
  formats: string;
  rating: number;
  nextSessionDate: string;
  duration: string;
  genres: string;
  description: string;
  releaseDate: string;
  venue?: string;
}): ShowItem {
  return {
    id: item.id,
    kind: item.kind,
    title: item.title,
    originalTitle: item.originalTitle,
    imageUrl: item.imageUrl,
    ageRating: item.ageRating,
    formats: item.formats,
    rating: item.rating,
    nextSessionDate: item.nextSessionDate,
    duration: item.duration,
    genres: item.genres,
    description: item.description,
    releaseDate: item.releaseDate,
    venue: item.venue,
  };
}

export function getShowItemById(id: string): ShowItem | undefined {
  const movie = getCatalogMovieById(id);
  if (movie) return movieToShowItem(movie);
  const event = getCatalogEventById(id);
  if (event) return eventToShowItem(event);
  const soon = getCatalogComingSoonById(id);
  if (soon) return comingSoonToShowItem(soon);
  return undefined;
}

export function getComingSoonShowItems(): ShowItem[] {
  return getCatalogComingSoon().map(comingSoonToShowItem);
}

export function getItemsByCategory(category: CategoryKey): ShowItem[] {
  if (category === 'cinema') return getCatalogMovies().map(movieToShowItem);
  if (category === 'home') {
    const topEvents = getCatalogEventsByCategory('home').map(eventToShowItem);
    const featuredMovies = getCatalogMovies().slice(0, 2).map(movieToShowItem);
    return [...topEvents, ...featuredMovies];
  }
  return getCatalogEventsByCategory(category).map(eventToShowItem);
}

File: backend/routers/catalog.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException
from sqlalchemy.orm import Session

from db import Event, get_db, json_loads

router = APIRouter(prefix="/api/catalog", tags=["catalog"])


def meta(row: Event) -> dict[str, Any]:
    return json_loads(row.movie_meta, {}) or {}


def event_to_catalog(row: Event) -> dict[str, Any]:
    m = meta(row)
    return {
        "id": row.id,
        "category": m.get("catalogCategory", row.category.lower()),
        "title": row.title,
        "venue": row.venue,
        "date": m.get("displayDate", row.date),
        "time": m.get("displayTime", row.time),
        "priceFrom": row.price_from,
        "priceTo": row.price_to,
        "imageUrl": row.image_url,
        "top": row.top,
        "description": row.description,
    }


def movie_to_catalog(row: Event) -> dict[str, Any]:
    m = meta(row)
    return {
        "id": row.id,
        "title": row.title,
        "originalTitle": m.get("originalTitle", row.title),
        "imageUrl": row.image_url,
        "ageRating": m.get("ageRating", "PG"),
        "formats": m.get("formats", "2D"),
        "rating": m.get("rating", 8.0),
        "nextSessionDate": m.get("nextSessionDate", ""),
        "duration": m.get("duration", ""),
        "genres": m.get("genres", ""),
        "description": row.description,
        "releaseDate": m.get("releaseDate", ""),
    }


def coming_soon_to_catalog(row: Event) -> dict[str, Any]:
    m = meta(row)
    return {
        "id": row.id,
        "kind": row.kind,
        "title": row.title,
        "originalTitle": m.get("originalTitle", row.title),
        "imageUrl": row.image_url,
        "ageRating": m.get("ageRating", "All Ages"),
        "formats": m.get("formats", ""),
        "rating": m.get("rating", 8.0),
        "nextSessionDate": m.get("nextSessionDate", ""),
        "duration": m.get("duration", ""),
        "genres": m.get("genres", ""),
        "description": row.description,
        "releaseDate": m.get("releaseDate", ""),
        "venue": m.get("venue", row.venue),
    }


@router.get("/events")
def list_events(category: str | None = None) -> list[dict[str, Any]]:
    with get_db() as db:
        rows = db.query(Event).filter(Event.is_coming_soon.is_(False), Event.kind != "movie").all()
        items = [event_to_catalog(r) for r in rows]
        if category == "home":
            return [i for i in items if i.get("top")]
        if category and category != "cinema":
            return [i for i in items if i["category"] == category]
        return items


@router.get("/movies")
def list_movies() -> list[dict[str, Any]]:
    with get_db() as db:
        rows = db.query(Event).filter(Event.kind == "movie", Event.is_coming_soon.is_(False)).all()
        return [movie_to_catalog(r) for r in rows]


@router.get("/coming-soon")
def list_coming_soon() -> list[dict[str, Any]]:
    with get_db() as db:
        rows = db.query(Event).filter(Event.is_coming_soon.is_(True)).all()
        return [coming_soon_to_catalog(r) for r in rows]


@router.get("/item/{item_id}")
def get_item(item_id: str) -> dict[str, Any]:
    with get_db() as db:
        row = db.query(Event).filter(Event.id == item_id).first()
        if not row:
            raise HTTPException(404, "Item not found")
        if row.kind == "movie":
            return {"kind": "movie", **movie_to_catalog(row)}
        if row.is_coming_soon:
            return {"kind": "coming_soon", **coming_soon_to_catalog(row)}
        return {"kind": "event", **event_to_catalog(row)}
